### ISSUE TITLE
Patching a StatefulSet's containerPorts causes values to be clobbered 

### ISSUE THREAD

**Issue by fabianvf** [kubernetes/kubernetes#64321]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
Created a StatefulSet with a container that had two ports, web-tcp (TCP, 80) and web-udp (UDP, 80) then patched the `spec.template.spec.containers.*.ports` without changing any values. This erased the web-tcp port and resulted in 2 copies of the web-udp port.

**What you expected to happen**:
Nothing, I would expect the ports to remain unchanged (as they do if I use kubectl apply instead).

**How to reproduce it (as minimally and precisely as possible)**:
```bash
cat <<EOF >test-sts.yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: statefulset-test
  namespace: default
spec:
  selector:
    matchLabels:
      app: nginx
  serviceName: "nginx"
  replicas: 1
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
          protocol: TCP
          name: web-tcp
        - containerPort: 80
          protocol: UDP
          name: web-udp
EOF

kubectl create -f test-sts.yaml
# At this point the ports match test-sts.yaml

kubectl patch -f test-sts.yaml -p '{"spec": {"template": {"spec": {"containers": [{"name": "nginx", "ports": [{"containerPort": 80, "protocol": "TCP", "name": "web-tcp"}, {"containerPort": 80, "protocol": "UDP", "name": "web-udp"}]}]}}}}'
# Now the TCP port has been clobbered and the port list contains two copies of the web-udp port
```

**Anything else we need to know?**:
[This was initially filed in the Ansible project against the k8s module](https://github.com/ansible/ansible/issues/40654)

The Ansible module uses the python client, which hits the REST API directly, so this is not an issue with kubectl.

Patching a Deployment in the same way results in an error because duplicate ports are not allowed - I assume it is doing the same incorrect merge but then failing validation, which is better at least.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
client Version: version.Info{Major:"", Minor:"", GitVersion:"v1.9.1+a0ce1bc657", GitCommit:"a0ce1bc", GitTreeState:"clean", BuildDate:"2018-04-12T16:13:45Z", GoVersion:"go1.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1+a0ce1bc657", GitCommit:"a0ce1bc", GitTreeState:"clean", BuildDate:"2018-05-16T21:26:57Z", GoVersion:"go1.9", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:
Running in a single node vagrant openshift environment with Intel Haswell Processor (16 cores, 128GB RAM)
- OS (e.g. from /etc/os-release): Fedora Atomic 27
- Kernel (e.g. `uname -a`):
```
Linux master.example.org 4.15.4-300.fc27.x86_64 #1 SMP Mon Feb 19 23:31:15 UTC 2018 x86_64 x86_64 x86_64 GNU/Linu
```
- Install tools: openshift-ansible

**Comment by fabianvf** [comment]
Seems to be related to https://github.com/kubernetes/kubernetes/issues/39188 , may be a dupe

**Comment by liggitt** [comment]
yes, dupe of #39188

/close

**Comment by fabianvf** [comment]
@liggit is the lack of validation for statefulsets included in #39188? (deployments at least fail out when this is hit, sts just breaks the ports list)

**Comment by liggitt** [comment]
statefulset validation is https://github.com/kubernetes/kubernetes/issues/60619
